Allow transformations on intermediate bindings streams produced during query execution

RubenEschauzier opened the pull request:
This PR aims to introduce a system that allows the creation of `process-iterator` actors that apply a transform on `streams` produced by actors subscribed to `bus-query-operation` and `bus-rdf-join`. Wrapper actors for `rdf-join` and `query-operation` apply the transform and can be enabled in the config.

An example usage pattern is in the PR. The `actor-process-iterator-record-intermediate-results` wraps around inner join streams and updates a `statistic-intermediate-results` with the produced intermediate result and any available metadata. Users can then hook into this statistic to determine what intermediate results are produced.

Currently, this PR is a draft because it has not been unit-tested. Initial experiments show it works for simple queries though. If possible, I would appreciate feedback on whether this is the correct way to implement this and if you have any high-level suggestions.

(Also configs are currently with it enabled, will update to disable by default).

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced multiple new configuration files for the Comunica framework, enhancing the capabilities of iterator transformations and query operations.
	- Added new actors and mediators to facilitate advanced RDF join operations and query processing, including `ActorQueryOperationWrapStream` and `ActorRdfJoinWrapStream`.
	- New `ActorIteratorTransformRecordIntermediateResults` class to update intermediate result statistics during transformations.
	- Added a new type definition `PartialResult` for flexible representation of partial results in RDF data handling.

- **Documentation**
	- Added a README file for the `@comunica/actor-process-iterator-record-intermediate-results` package, detailing installation and configuration.

- **Tests**
	- Comprehensive test suites added for various new and updated actors, ensuring functionality and reliability across different scenarios, including tests for `ActorRdfJoinWrapStream` and `ActorQueryOperationWrapStream`.

- **Chores**
	- New package.json files created for several packages to manage dependencies and metadata effectively.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rubensworks commented:
@coderabbitai review

rubensworks commented on packages/actor-query-operation-wrap-stream/lib/ActorQueryOperationWrapStream.ts:
```diff
@@ -0,0 +1,74 @@
+    const output: IQueryOperationResult = await this.mediatorQueryOperation.mediate(action);
+    switch (output.type) {
+      case 'bindings':
+        output.bindingsStream = <AsyncIterator<RDF.Bindings>>
```
I wonder if we also should allow the metadata to be optionally processed?

This could be useful if someone were to add a process iterator actor that does some filtering, and thereby could reduce the cardinality and update it in the metadata.

rubensworks commented on packages/bus-process-iterator/lib/ActorProcessIterator.ts:
```diff
@@ -0,0 +1,89 @@
+import type { types } from 'sparqlalgebrajs/lib/algebra';
+
+/**
+ * A comunica actor for process-iterator events.
```
Just thinking about naming.

When I read this, my first thought also goes to a process at OS-level. Like a Node.js process, or a Firefox process.
But maybe this is just me. But if you agree that naming could be improved, something like `transform-iterator` could be an alternative.

rubensworks reviewed (changes requested):
The overall approach looks good to me!

RubenEschauzier commented:
Thanks for the review! I changed the name and added the metadata field to the iterator-transform bus. I'll start writing unit tests to finish the PR and ping you when it's ready for another (more in-depth) review.

rubensworks commented on engines/config-query-sparql/config/process-iterator/actors.json:
```diff
@@ -0,0 +1,15 @@
+  ],
+  "@id": "urn:comunica:default:Runner",
+  "@type": "Runner",
+  "actors": [
+    {
+      "@id": "urn:comunica:default:iterator-transform/actors#record-intermediate-results",
+      "@type": "ActorIteratorTransformRecordIntermediateResults",
+      "wraps": ["inner", "minus", "optional", "project"]
+    }
+  ]
```
No, what @RubenEschauzier did seems fine to me.

rubensworks commented on engines/config-query-sparql/config/query-operation/actors/query/wrap-stream.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "@context": [
+    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/runner/^3.0.0/components/context.jsonld",
+
+    "https://linkedsoftwaredependencies.org/bundles/npm/@comunica/actor-query-operation-wrap-stream/^3.0.0/components/context.jsonld"
+  ],
```
No, this seems fine.

rubensworks commented on packages/bus-query-operation/lib/ActorQueryOperation.ts:
```diff
@@ -255,6 +255,10 @@ export abstract class ActorQueryOperation extends Actor<IActionQueryOperation, I
     return shape.operation.pattern.type === operation.type;
   }
+
+  public static setContextWrapped(context: IActionContext, value: boolean): IActionContext {
```
I understand the need for calling this function.
However, it feels error-prone, as forgetting to call it may lead to difficult bugs that are difficult to debug.

An alternative approach could be to make `KEY_CONTEXT_WRAPPED_QUERY_OPERATION` refer not to a boolean, but the actual operation object that is being wrapped. Then you can just check in the actor's test if the wrapped operation is the same as current or not to decide if the test should pass.

rubensworks commented on packages/actor-query-operation-wrap-stream/README.md:
```diff
@@ -0,0 +1,42 @@
+[![npm version](https://badge.fury.io/js/%40comunica%2Factor-query-operation-wrap-stream.svg)](https://www.npmjs.com/package/@comunica/actor-query-operation-wrap-stream)
+
+A comunica Wrap Stream Query Operation Actor. This actor calls the `bus-iterator-transform` for all iterators produced by `query-operation` actors.
+This actor should only be included if you require the functionality of `bus-iterator-transform` as it will likely slow down query execution.
```
```suggestion
This actor should only be included if you require the functionality of `bus-iterator-transform` as it may slow down query execution.
```

rubensworks commented on packages/actor-query-operation-wrap-stream/lib/ActorQueryOperationWrapStream.ts:
```diff
@@ -0,0 +1,93 @@
+            context: action.context,
+            metadata: {
+              type: output.type,
+              ...await output.metadata(),
```
Can't we just pass `output.metadata` directly without the `await`?

RubenEschauzier replied:
I don't think so, as the metadata function returns a promise.

rubensworks commented on packages/actor-query-operation-wrap-stream/lib/ActorQueryOperationWrapStream.ts:
```diff
@@ -0,0 +1,93 @@
+        output.metadata = <() => Promise<MetadataQuads>> iteratorTransformed.streamMetadata;
+        break;
+      }
+      default: {
+        break;
+      }
+    }
```
No, the current approach is good.

rubensworks commented on packages/actor-rdf-join-wrap-stream/README.md:
```diff
@@ -0,0 +1,40 @@
+[![npm version](https://badge.fury.io/js/%40comunica%2Factor-rdf-join-wrap-stream.svg)](https://www.npmjs.com/package/@comunica/actor-rdf-join-wrap-stream)
+
+A comunica Wrap Stream RDF Join Actor. This actor calls the `bus-iterator-transform` for all iterators produced by `query-operation` actors.
+This actor should only be included if you require the functionality of `bus-iterator-transform` as it will likely slow down query execution.
```
```suggestion
This actor should only be included if you require the functionality of `bus-iterator-transform` as it may slow down query execution.
```

rubensworks commented on packages/actor-rdf-join-wrap-stream/lib/ActorRdfJoinWrapStream.ts:
```diff
@@ -0,0 +1,90 @@
+    action.context = ActorRdfJoin.setContextWrapped(action.context, true);
+    const result: IQueryOperationResultBindings = await this.mediatorJoin.mediate(action);
+
+    const { stream, streamMetadata } = (await this.mediatorIteratorTransform.mediate(
+      {
+        operation: action.type,
+        stream: result.bindingsStream,
+        streamMetadata: result.metadata,
+        context: action.context,
+        metadata: {
+          joinEntries: action.entries.length,
+          ...await result.metadata(),
+          resultContext: result.context,
+        },
+      },
+    ));
```
Not needed.

rubensworks commented on packages/actor-rdf-join-wrap-stream/lib/ActorRdfJoinWrapStream.ts:
```diff
@@ -0,0 +1,90 @@
+      {
+        operation: action.type,
+        stream: result.bindingsStream,
+        streamMetadata: result.metadata,
```
Why both metadata and streamMetadata?

RubenEschauzier replied:
The intention is to allow users to change the `streamMetadata` (the metadata function given by the join result) in iterator transform actors. The `metadata` field is intended as general metadata including more than just the stream metadata. If that makes sense.

rubensworks replied:
I may be misunderstanding something, but if users want to change the metadata, this could be done in the return value of the mediator, right?
I suspect we may only need `streamMetadata`.

RubenEschauzier replied:
I think the intention was to allow actors on `bus-transform-iterator` to apply a transformation on the intermediate iterator's `metadata` function. For example, if you filter some results and want to update the estimated cardinality. In this case `streamMetadata` is the `metadata` function of the intermediate iterator (which may be somewhat confusingly named).

rubensworks replied:
Ok, but then `streamMetadata` is all we need, and `metadata` can be removed (and perhaps then `streamMetadata` can be renamed to `metadata`)?

RubenEschauzier replied:
But some metadata could be useful to pass through to the `iterator-transform` actors that are not in `streamMetadata`, like in `rdf-join-wrap-stream`, the number of entries in the wrapped action shows how many joins were already performed. Where would you suggest this be added?

rubensworks replied:
Aha, now I understand you.
I didn't understand the need for two types of metadata fields, but now I do.

Instead of passing custom fields such as the number of join entries in `metadata`, why not omit this custom metadata, and instead directly pass the original `action` via e.g. `originalAction`, to allow the user to derive the number of join entries like that?

RubenEschauzier replied:
Right, that works!

rubensworks commented on packages/bus-iterator-transform/lib/ActorIteratorTransform.ts:
```diff
@@ -0,0 +1,133 @@
+}
+
+/**
+ * TODO If this fails for multiple transforms its because of lack of context, SO CHECK IT
```
TODO

rubensworks commented on packages/context-entries/lib/Keys.ts:
```diff
@@ -233,6 +234,12 @@ export const KeysQueryOperation = {
+   * The process-iterator mediator that will wrap operation outputs of query-operation and rdf-join actors
+   * to allow users to define processors over Asynciterators
+   */
+  mediatorProcessIterator: new ActionContextKey<any>
```
This one is not needed (anymore)?

rubensworks commented on packages/statistic-intermediate-results/lib/StatisticIntermediateResults.ts:
```diff
@@ -0,0 +1,19 @@
+    this.key = KeysStatistics.intermediateResults;
+  }
+
+  public updateStatistic(intermediateResult: IPartialResult): boolean {
+    intermediateResult.metadata.time = performance.now();
+    this.emit(intermediateResult);
+    return true;
+  }
```
No.

rubensworks commented on packages/types/lib/statistics/IPartialResult.ts:
```diff
@@ -0,0 +1,6 @@
+import type * as RDF from '@rdfjs/types';
+
+export interface IPartialResult {
+  data: RDF.Bindings | RDF.Quad;
+  metadata: Record<string, any>;
```
No.

rubensworks commented on packages/types/lib/statistics/IPartialResult.ts:
```diff
@@ -0,0 +1,6 @@
+import type * as RDF from '@rdfjs/types';
+
+export interface IPartialResult {
+  data: RDF.Bindings | RDF.Quad;
```
To enable more easy type-checking when using these partial results, I propose something like this:

```
export type PartialResult = {
  type: 'bindings';
  data: RDF.Bindings;
} | { type: 'quad'; ... };
```

RubenEschauzier replied:
So I checked to see if this works, but the problem is that when you use such an approach you have to determine when your input is a `binding` or `quad`. I could make a mapping from operation to quad or binding but that seems like a lot of unnecessary work. How do you see this work?

rubensworks replied:
It can be used as follows:
```
const partialResult: PartialResult = ...;
if (partialResult.type === 'bindings') {
  // process data as bindings
}
```

rubensworks replied:
Or are you referring to using this interface in your transformer?

Perhaps we need different transform operations for the different types to simplify handling?

RubenEschauzier replied:
I was referring to 'creating' the input to `updateStatistic`, but I see now that it shouldn't be too difficult.

RubenEschauzier replied:
I've changed the typings around (a lot) following advice from Joachim. I'm not sure the typings are better this way so please let me know what you think!

rubensworks reviewed (changes requested):
Good stuff!

I closed the bot's comments I disagree with.

RubenEschauzier commented on packages/statistic-intermediate-results/README.md:
````diff
@@ -0,0 +1,45 @@
+This class must be added to the context with the key defined in the class.
+
+This module is part of the [Comunica framework](https://github.com/comunica/comunica),
+
+[Click here if you just want to query with Comunica](https://comunica.dev/docs/query/).
+
+## Install
+
+```bash
+$ yarn add @comunica/statistic-link-dereference
+```
+## Using the tracker
+
+### Adding the tracker to the context
+```javascript
+    const QueryEngine = require('@comunica/query-sparql').QueryEngine;
+    const myEngine = new QueryEngine();
+    const statisticTracker = new StatisticIntermediateResults();
+
+    // Print any data emitted by the tracker
+    statisticTracker.on((intermediateResult) => console.log(intermediateResult));
+
+    // Add the tracker to the context
+    let context = {sources: ['https://fragments.dbpedia.org/2015/en']};
+    context[statisticTracker.key.name] = statisticTracker;
+
+    // Execute the query
+    const bindingsStream = await myEngine.queryBindings(`
+    SELECT ?s ?p ?o WHERE {
+        ?s ?p ?o
+    } LIMIT 10`, context);
+    bindingsStream.on('data', (binding) => {
+        console.log(binding.toString());
+    });
+```
+
+### Allowing Comunica to call the tracker
+Comunica does not call the `updateStatistic` function by default. To enable this users should first add the following actors to the config:
+`actor-query-operation-wrap-stream` and `actor-rdf-join-wrap-stream`. This will call the `bus-iterator-transform` for each intermediate iterator produced
+by Comunica. Users should then add `actor-iterator-transform-record-intermediate-results` to the config, which will call the `updateStatistic` method of `statistic-intermediate-result` for each produced intermediate result.
````
Do we need the extra fields? I can understand the usefulness of writing down the output format, although that will be pretty long and depend on the intermediate result type. But the last three don't seem useful.

rubensworks replied:
Yeah, let's not do that here. Maybe something for the website docs in the future.

RubenEschauzier commented on packages/bus-query-operation/test/ActorQueryOperationTyped-test.ts:
```diff
@@ -81,6 +81,7 @@ describe('ActorQueryOperationTyped', () => {
         [KeysInitQuery.physicalQueryPlanLogger.name]: logger,
         [KeysInitQuery.physicalQueryPlanNode.name]: operation,
         [KeysQueryOperation.operation.name]: operation,
+        [KEY_CONTEXT_WRAPPED_QUERY_OPERATION.name]: false,
```
I don't think the new key should be added here, as it is only added during query execution or does not interact with these tests.

RubenEschauzier commented on packages/actor-query-operation-update-load/test/ActorQueryOperationLoad-test.ts:
```diff
@@ -115,6 +116,7 @@ describe('ActorQueryOperationLoad', () => {
         operation: expect.anything(),
         context: new ActionContext({
           '@comunica/bus-query-operation:operation': expect.anything(),
+          [KEY_CONTEXT_WRAPPED_QUERY_OPERATION.name]: false,
```
This would probably make things more complex

RubenEschauzier commented on packages/bus-iterator-transform/lib/ActorIteratorTransform.ts:
```diff
@@ -0,0 +1,133 @@
+  /**
+   * @param args - @defaultNested {<default_bus> a <cc:components/Bus.jsonld#Bus>} bus
+   */
+  public constructor(args: IActorIteratorTransformArgs) {
+    super(args);
+  }
```
Are we sure this should be removed? Other actors keep this 'useless constructor' too

rubensworks replied:
Yeah, the ctor must remain in this case.

rubensworks commented:
@RubenEschauzier Before I forget; since we want to include this in the upcoming major release, could you retarget this PR to `next/major`?

rubensworks commented on packages/actor-iterator-transform-record-intermediate-results/lib/ActorIteratorTransformRecordIntermediateResults.ts:
```diff
@@ -0,0 +1,52 @@
+    MetadataBindings | MetadataQuads
+> {
+
+  public async transformIterator(action: IActionIteratorTransformQuad): 
```
I'm not sure I understand the need for these method overrides.
AFAICS, they are not actually being used anymore?
If so, the signature from line 32 should be sufficient I guess?

RubenEschauzier commented on packages/actor-iterator-transform-record-intermediate-results/lib/ActorIteratorTransformRecordIntermediateResults.ts:
```diff
@@ -0,0 +1,57 @@
+export class ActorIteratorTransformRecordIntermediateResults extends ActorIteratorTransform {
+  public async transformIteratorBindings(action: IActionIteratorTransformBindings):
+  Promise<ITransformIteratorOutput<AsyncIterator<RDF.Bindings>, MetadataBindings>> {
+    const statisticIntermediateResults: StatisticIntermediateResults = action.context
+      .getSafe(KeysStatistics.intermediateResults);
+    const output = action.stream.map((data) => {
+      statisticIntermediateResults.updateStatistic(
+        {
+          type: <'bindings'> action.type,
+          data,
+          metadata: {
+            operation: action.operation,
+            ...action.metadata,
+          },
+        },
+      );
+      return data;
+    });
+    return { stream: output, metadata: <() => Promise<MetadataBindings>> action.metadata };
+  }
+
+  public async transformIteratorQuad(action: IActionIteratorTransformQuad):
+  Promise<ITransformIteratorOutput<AsyncIterator<RDF.Quad>, MetadataQuads>> {
+    const statisticIntermediateResults: StatisticIntermediateResults = action.context
+      .getSafe(KeysStatistics.intermediateResults);
+    const output = action.stream.map((data) => {
+      statisticIntermediateResults.updateStatistic(
+        {
+          type: <'quad'> action.type,
+          data,
+          metadata: {
+            operation: action.operation,
+            ...action.metadata,
+          },
+        },
+      );
+      return data;
+    });
+    return { stream: output, metadata: <() => Promise<MetadataQuads>> action.metadata };
+  }
```
This does not work.

RubenEschauzier commented on packages/bus-iterator-transform/lib/ActorIteratorTransform.ts:
```diff
@@ -0,0 +1,171 @@
+    super(args);
+  }
+
+  public async run(action: ActionIteratorTransform):
+  Promise<ActorIteratorTransformOutput> {
+    if (action.type === 'bindings') {
+      const { stream, metadata } = await this.transformIteratorBindings(action);
+      return {
+        type: action.type,
+        operation: action.operation,
+        stream,
+        metadata,
+        originalAction: action.originalAction,
+        context: action.context,
+      };
+    }
+
+    const { stream, metadata } = await this.transformIteratorQuad(action);
+    return {
+      type: action.type,
+      operation: action.operation,
+      stream,
+      metadata,
+      originalAction: action.originalAction,
+      context: action.context,
+    };
+  }
```
This also does not work

RubenEschauzier commented:
I updated the code to a ready state. The tests still fail due to changed function names and types, but this will be resolved once the general approach seems good.
I also changed some configs for local testing, this will also be reverted once the approach is good to go.

rubensworks commented on packages/actor-query-source-identify-hypermedia-annotate-source/test/ActorQuerySourceIdentifyHypermediaAnnotateSource-test.ts:
```diff
@@ -86,7 +86,7 @@ describe('ActorQuerySourceIdentifyHypermediaSourceAttribution', () => {
         quad('s2', 'p2', 'o2'),
       ]);
-      context = context.set(KEY_CONTEXT_WRAPPED, true);
+      context = context.set(KEY_CONTEXT_WRAPPED, actor);
```
This looks strange; is it correct?

rubensworks commented on packages/actor-rdf-join-wrap-stream/lib/ActorRdfJoinWrapStream.ts:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  public override async test(action: IActionRdfJoin): Promise<IMediatorTypeJoinCoefficients> {
+    if (action.context.get(KEY_CONTEXT_WRAPPED_RDF_JOIN) === this) {
```
I would advise against storing the actor inside the context.
Instead, you can store the operation that is being wrapped, and test based on that.

rubensworks commented on packages/bus-iterator-transform/lib/ActorIteratorTransform.ts:
```diff
@@ -0,0 +1,170 @@
+  public async test(
+    action: ActionIteratorTransform,
+  ): Promise<IActorTest> {
+    if (!(this.wraps === undefined) && !this.wraps.includes(action.operation)) {
```
I think something like this will also work:
```suggestion
    if (!this.wraps?.includes(action.operation)) {
```

RubenEschauzier replied:
It doesn't seem so, it returns true when `this.wraps` is undefined, while that should return false, as `this.wraps === undefined` should make the actor wrap all operations.

rubensworks replied:
The explicit `undefined` check can surely go, or be replaced by a falsy check.

RubenEschauzier commented on packages/actor-iterator-transform-record-intermediate-results/lib/ActorIteratorTransformRecordIntermediateResults.ts:
```diff
@@ -0,0 +1,62 @@
+    super(args);
+  }
+
+  public async transformIteratorBindings(action: IActionIteratorTransformBindings):
+  Promise<ITransformIteratorOutput<AsyncIterator<RDF.Bindings>, MetadataBindings>> {
+    const statisticIntermediateResults: StatisticIntermediateResults = action.context
+      .getSafe(KeysStatistics.intermediateResults);
+    const output = action.stream.map((data) => {
+      statisticIntermediateResults.updateStatistic(
+        {
+          type: action.type,
+          data,
+          metadata: {
+            operation: action.operation,
+            metadata: action.metadata,
+          },
+        },
+      );
+      return data;
+    });
+    return { stream: output, metadata: action.metadata };
+  }
```
@rubensworks I'm not sure about this one. Should we leave the `metadata` call to the statistic or do it in the `transformIterator` call? I'm leaning towards calling it in the statistic, as statistics might not use the metadata so it would increase overhead for no reason.

rubensworks replied:
I agree with your opinion @RubenEschauzier, let's leave calling it to the statistic itself.

rubensworks commented on packages/actor-iterator-transform-record-intermediate-results/README.md:
```diff
@@ -0,0 +1,40 @@
+# Comunica Count Intermediate Results Process Iterator Actor
```
The `.npmignore` files can be removed since v4.

rubensworks commented on packages/actor-iterator-transform-record-intermediate-results/package.json:
```diff
@@ -0,0 +1,47 @@
+  "description": "A record-intermediate-results iterator-transform actor",
+  "lsd:module": true,
+  "license": "MIT",
+  "homepage": "https://comunica.dev/",
```
Can you add the `funding` entry to all package.json's which are now present in all packages since v4?

rubensworks commented on packages/actor-query-operation-source/lib/ActorQueryOperationSource.ts:
```diff
@@ -44,8 +44,8 @@ export class ActorQueryOperationSource extends ActorQueryOperation {
-    const mergedContext = sourceWrapper.context ? action.context.merge(sourceWrapper.context) : action.context;
-
+    let mergedContext = sourceWrapper.context ? action.context.merge(sourceWrapper.context) : action.context;
+    mergedContext = setContextWrapped(action.operation, mergedContext);
```
Are you sure this is needed here?
I would expect this context entry to only have to be set within the actor doing the actual wrapping.

rubensworks commented on packages/bus-query-operation/lib/ActorQueryOperationTyped.ts:
```diff
@@ -56,6 +57,8 @@ TS = undefined,
       action.context = action.context.set(KeysInitQuery.physicalQueryPlanNode, action.operation);
     }
+    // Set wrapped to false to allow recursive calls to query operation to also be wrapped
+    action.context = setContextWrapped(action.operation, action.context);
```
I suspect this also to not be required.

rubensworks commented on packages/bus-rdf-join/lib/ActorRdfJoin.ts:
```diff
@@ -433,6 +433,8 @@ TS
       );
     }
 
+    action.context = this.setContextWrapped(action, action.context);
```
Same here.

rubensworks commented on packages/bus-rdf-join/lib/ActorRdfJoin.ts:
```diff
@@ -544,11 +556,6 @@ export interface IActorRdfJoinInternalOptions {
-/**
- * Represents a logical join type.
- */
-export type LogicalJoinType = 'inner' | 'optional' | 'minus';
```
Be careful: this will be a breaking change.

RubenEschauzier replied:
Right, I added the following line: `// TODO remove this in next major version
export { LogicalJoinType } from '@comunica/types';`.
Is that ok? I figured moving it would be better as it it also used by the `iterator-transform` actors.

rubensworks replied:
Yep, something like that is ok!
I don't see it in the diff yet though, so I assume it still needs to be pushed?

RubenEschauzier replied:
I added it at line 593, is that ok?

RubenEschauzier replied:
It is [here](https://github.com/comunica/comunica/blob/332ad8b1ce2f9ccf5f9ff1fb24815b88b4b1356c/packages/bus-rdf-join/lib/ActorRdfJoin.ts#L593), not sure how to post the permalink to the diff, but it does show up for me.

rubensworks replied:
Looks perfect!

rubensworks commented on packages/types/lib/statistics/IPartialResult.ts:
```diff
@@ -0,0 +1,7 @@
+  type: 'bindings';
+  data: RDF.Bindings;
+  metadata: Record<string, any>;
+} | { type: 'quads'; data: RDF.Quad; metadata: Record<string, any> };
```
Let's put these on separate lines as well.

rubensworks commented on packages/utils-query-operation/lib/Utils.ts:
```diff
@@ -110,3 +110,20 @@ export function removeOperationSource(operation: Algebra.Operation): void {
+ * @param context The current action context.
+ * @returns A new action context with the operation marked as wrapped.
+ */
+export function setContextWrapped(operation: Algebra.Operation, context: IActionContext): IActionContext {
```
Just a question, as you probably have a better view on this than me:
Is this util function (and context key) required here?
Can we not move it to the wrapper actor itself?
Because I suspect that only that wrapper actor will ever need to call this and use this context key?

The same applies to the util function in `ActorRdfJoin`.

RubenEschauzier replied:
Yep, I agree.

rubensworks reviewed (changes requested):
Just a minor comment.
Other than that, this looks good to go!

Just double-checking; this is fully backwards-compatible, right?
(AFAICS, it is, since only new actors are added, and the new actors are not added to the default config)

rubensworks commented:
@RubenEschauzier Is anything else needed for this PR? (also see my last question above regarding backwards-compat)

RubenEschauzier commented:
It should be backward compatible. The only place I've changed the existing code is in `packages/context-entries/lib/Keys.ts`, where I've changed the underlying name of a key. Now that I think of it, if someone were to rely on the name of the key instead of the object, this could cause issues. But that is an edge case.

rubensworks commented:
Very nice work @RubenEschauzier!
Do make sure to document how to use this on the website in the near future.
We'll probably need a new page dedicated to this in the advanced modification docs.
